feat: Adding support for multiple languages

jhammond2012 opened the pull request:
## Description

The existing documentation only supports a single language. The goal here is to set up support for multiple languages. We experienced issues trying to use the 'out of the box' i18n support but due to how the SSR is set up for the MD files, we had to change the way the route was generated. For this, we created a middleware here: 

`src/pages/_middleware.ts`

This checks for the browser's locale and compares it to the supported languages defined in `src/common/constants.ts` via `SUPPORTED_LANGUAGES`. If we support the language, a redirect happens and the correct MD files are rendered. You can find an un-translated example here: `src/pages/es`. These are copies of the existing MD files and will be translated by the community.

## Checklist

- [x] [Conventional commits were used](https://www.conventionalcommits.org/en/v1.0.0/)
- [x] New links to files and images were verified
- [x] For fixes/refactors: all existing references were identified and replaced
- [x] [Style guide](https://developers.google.com/style) was reviewed and applied
- [x] Clear code samples were provided
- [x] People were tagged for review

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/blockstack/docs?pullRequest=1283) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/blockstack/docs?pullRequest=1283) before we can accept your contribution.<br/>**1** out of **2** committers have signed the CLA.<br/><br/>:white_check_mark: jhammond2012<br/>:x: anilkumar1994<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/blockstack/docs?pullRequest=1283) it.</sub>

jhammond2012 commented:
@alexkeating Sorry for all of the "Chinese Simplified" commits. I was testing a tool for translation collab and it made a bunch of commits. I have removed the file but the commits stay in the history.

agraebe commented:
> @agraebe Sorry for all of the "Chinese Simplified" commits. I was testing a tool for translation collab and it made a bunch of commits. I have removed the file but the commits stay in the history.

conventional commit usage would be great. any chance you could rebase the 97 commits to improve the git history?

did you also check why it doesn't deploy? is there some build issue or similar?

jhammond2012 commented:
> > @agraebe Sorry for all of the "Chinese Simplified" commits. I was testing a tool for translation collab and it made a bunch of commits. I have removed the file but the commits stay in the history.
> 
> conventional commit usage would be great. any chance you could rebase the 97 commits to improve the git history?
> 
> did you also check why it doesn't deploy a preview? there seems to be a build issue:
> 
> ```
> Failed to compile.
> 
> ./src/pages/_middleware.ts:1:43
> Type error: Cannot find module 'next/server' or its corresponding type declarations.
> 
> > 1 | import { NextRequest, NextResponse } from 'next/server';
>     |                                           ^
>   2 | import { SUPPORTED_LANGUAGES } from '../common/constants';
>   3 | 
>   4 | export function middleware(request: NextRequest) {
> ```

Is there a way for me to get the latest build errors from Vercel? Thanks!

agraebe commented:
I usually just run 'yarn build' locally. That's what's done on Vercel for the deployment action

jhammond2012 commented:
@agraebe To get the next module working that was needed, I had to upgrade to the next version. This caused some new errors from eslint. I fixed a couple but some of them are going to require a larger overtaking. In some errors, I included an eslint disable for the line affected. This does not cause any application errors or build errors but just conflicts with the eslint rules. Should we keep this PR open until all of those are updated or can that be a separate PR?

Thanks!

agraebe commented:
can you temporarily disable the eslint rule(s) so they won't throw all these errors? re-enabling and fixing the issues could be part of a new PR IMHO.

Who and how should test the outcomes? it's a lot of files, so I'm not sure how to best review. I did notice a typo on the landing page (`Learn about 1234 Stacks mining`). Are the URLs supposed to include the language used? How would one switch?

jhammond2012 commented:
@agraebe I have disabled the rules `require-await` and `no-var-requires`. Additionally, I have removed the /es folder entirely and removed it from the list of supported languages since we do not have them translated yet. I should have pulled them out originally since this was only meant to add support for multiple languages and not the translations themselves.

agraebe reviewed (approved):
lgtm

jhammond2012 commented:
@agraebe will this code_quality task eventually run or should I do something else?

friedger reviewed (approved):
The code changes look reasonable. Not an expert here.

jhammond2012 commented on src/common/routes/get-routes.js:
```diff
@@ -63,9 +65,11 @@ const getHeadings = mdContent => {
   try {
     const fileContent = fs.readFileSync(
-      path.join('./src/pages', `${route === '/' ? 'index' : route}.md`),
+      path.join('./src/pages', `${route === '/' ? 'index' : route}/index.md`),
```
Having the language files in a folder helped the organizational efforts with the integration (https://crowdin.com/). This change was not 'required' from a technical standpoint but rather how the platform rendered the translations.

agraebe commented:
> @agraebe will this code_quality task eventually run or should I do something else?

I just initiated the checks 👍🏼

jhammond2012 commented:
> > @agraebe will this code_quality task eventually run or should I do something else?
> 
> I just initiated the checks 👍🏼

Oh awesome, thanks.

friedger commented:
Great work!
